README.rst
========
Overview
========

.. start-badges

.. list-table::
    :stub-columns: 1

    * - docs
      - |docs|
    * - tests
      - | |travis|
        |
    * - package
      - | |version| |wheel| |supported-versions| |supported-implementations|
        | |commits-since|
.. |docs| image:: https://readthedocs.org/projects/python-lmt8x/badge/?style=flat
    :target: https://python-lmt8x.readthedocs.io/
    :alt: Documentation Status

.. |travis| image:: https://api.travis-ci.com/charlee/python-lmt8x.svg?branch=master
    :alt: Travis-CI Build Status
    :target: https://travis-ci.com/github/charlee/python-lmt8x

.. |version| image:: https://img.shields.io/pypi/v/lmt8x.svg
    :alt: PyPI Package latest release
    :target: https://pypi.org/project/lmt8x

.. |wheel| image:: https://img.shields.io/pypi/wheel/lmt8x.svg
    :alt: PyPI Wheel
    :target: https://pypi.org/project/lmt8x

.. |supported-versions| image:: https://img.shields.io/pypi/pyversions/lmt8x.svg
    :alt: Supported versions
    :target: https://pypi.org/project/lmt8x

.. |supported-implementations| image:: https://img.shields.io/pypi/implementation/lmt8x.svg
    :alt: Supported implementations
    :target: https://pypi.org/project/lmt8x

.. |commits-since| image:: https://img.shields.io/github/commits-since/charlee/python-lmt8x/v1.0.0.svg
    :alt: Commits since latest release
    :target: https://github.com/charlee/python-lmt8x/compare/v1.0.0...master



.. end-badges

Transfer tables for LMT8x Temperature Sensors.

LMT8x (LMT84, LMT85, LMT86, LMT87) are a series of `analog temperature sensors <https://www.ti.com/sensors/temperature-sensors/overview.html>`_ made by Texus Instrument.
These sensors can provide -50°C ~ 150°C with ±0.4°C accuracy. Although the output voltage is nearly linear to the temperature, it does have a slight umbrella parabolic shape.
Therefore transfer tables are required to convert the voltage to the temperature.

Datasheets:

- `LMT84 <https://www.ti.com/lit/ds/symlink/lmt84.pdf>`_
- `LMT85 <https://www.ti.com/lit/ds/symlink/lmt85.pdf>`_
- `LMT86 <https://www.ti.com/lit/ds/symlink/lmt86.pdf>`_
- `LMT87 <https://www.ti.com/lit/ds/symlink/lmt87.pdf>`_


This package provides tranfer functions for LMT84, LMT85, LMT86, and LMT87 based on their transfer tables.
These funcitons basically do a binary search through the transfer tables and return the match.
If no match found, linear interpolation will be used to generate an approximate value.


Installation
============

::

    pip install lmt8x

You can also install the in-development version with::

    pip install https://github.com/charlee/python-lmt8x/archive/master.zip

Usage
======

`lmt8x` package provides functions `lmt84_v2t`, `lmt85_v2t`, `lmt86_v2t`, and `lmt87_v2t` for converting voltage to temperature.

- input voltage must be in mV.
- output temperature is in celsius.

::

  from lmt8x import lmt87_v2t     # or `lmt86_v2t`, `lmt85_v2t`, `lmt84_v2t`

  # read voltage from sensors.
  # v = read_sensor()

  # convert v to temperature.
  # the pamaeter must be in mV. Return value is in celsius.
  temp = lmt87_v2t(v * 1000)

  print('Temperature is %s C.' % temp)
  

Documentation
=============


https://python-lmt8x.readthedocs.io/


Development
===========

To run all the tests run::

    tox

Note, to combine the coverage data from all the tox environments run:

.. list-table::
    :widths: 10 90
    :stub-columns: 1

    - - Windows
      - ::

            set PYTEST_ADDOPTS=--cov-append
            tox

    - - Other
      - ::

            PYTEST_ADDOPTS=--cov-append tox


CHANGELOG.rst

Changelog
=========

1.0.0 (2021-04-02)
------------------

* First release on PyPI.


tests/test_lmt8x.py
import unittest

from lmt8x import lmt84_v2t, lmt85_v2t, lmt86_v2t, lmt87_v2t


class LMT84TestCase(unittest.TestCase):

    def test_exact_temp(self):
        self.assertEqual(lmt84_v2t(1299), -50)
        self.assertEqual(lmt84_v2t(1294), -49)
        self.assertEqual(lmt84_v2t(1034), 0)
        self.assertEqual(lmt84_v2t(183), 150)

    def test_linear_interpolation(self):
        self.assertGreater(lmt84_v2t(1298), -50)
        self.assertLess(lmt84_v2t(1298), -49)

    def test_invalid_values(self):
        with self.assertRaises(ValueError):
            lmt84_v2t(1500)

        with self.assertRaises(ValueError):
            lmt84_v2t(0)


class LMT85TestCase(unittest.TestCase):

    def test_exact_temp(self):
        self.assertEqual(lmt85_v2t(1955), -50)
        self.assertEqual(lmt85_v2t(1949), -49)
        self.assertEqual(lmt85_v2t(1567), 0)
        self.assertEqual(lmt85_v2t(301), 150)

    def test_linear_interpolation(self):
        self.assertGreater(lmt85_v2t(1954), -50)
        self.assertLess(lmt85_v2t(1954), -49)

    def test_invalid_values(self):
        with self.assertRaises(ValueError):
            lmt85_v2t(2000)

        with self.assertRaises(ValueError):
            lmt85_v2t(0)


class LMT86TestCase(unittest.TestCase):

    def test_exact_temp(self):
        self.assertEqual(lmt86_v2t(2616), -50)
        self.assertEqual(lmt86_v2t(2607), -49)
        self.assertEqual(lmt86_v2t(2100), 0)
        self.assertEqual(lmt86_v2t(420), 150)

    def test_linear_interpolation(self):
        self.assertGreater(lmt86_v2t(2615), -50)
        self.assertLess(lmt86_v2t(2615), -49)

    def test_invalid_values(self):
        with self.assertRaises(ValueError):
            lmt86_v2t(3000)

        with self.assertRaises(ValueError):
            lmt86_v2t(0)


class LMT87TestCase(unittest.TestCase):

    def test_exact_temp(self):
        self.assertEqual(lmt87_v2t(3277), -50)
        self.assertEqual(lmt87_v2t(3266), -49)
        self.assertEqual(lmt87_v2t(2633), 0)
        self.assertEqual(lmt87_v2t(538), 150)

    def test_linear_interpolation(self):
        self.assertGreater(lmt87_v2t(3276), -50)
        self.assertLess(lmt87_v2t(3276), -49)

    def test_invalid_values(self):
        with self.assertRaises(ValueError):
            lmt87_v2t(4000)

        with self.assertRaises(ValueError):
            lmt87_v2t(0)


src/lmt8x/lmt84.py
# datasheet: https://www.ti.com/lit/ds/symlink/lmt84.pdf
transfer_table = {
    'range': [-50, 150],
    'vout': [
        1299,
        1294,
        1289,
        1284,
        1278,
        1273,
        1268,
        1263,
        1257,
        1252,
        1247,
        1242,
        1236,
        1231,
        1226,
        1221,
        1215,
        1210,
        1205,
        1200,
        1194,
        1189,
        1184,
        1178,
        1173,
        1168,
        1162,
        1157,
        1152,
        1146,
        1141,
        1136,
        1130,
        1125,
        1120,
        1114,
        1109,
        1104,
        1098,
        1093,
        1088,
        1082,
        1077,
        1072,
        1066,
        1061,
        1055,
        1050,
        1044,
        1039,
        1034,
        1028,
        1023,
        1017,
        1012,
        1007,
        1001,
        996,
        990,
        985,
        980,
        974,
        969,
        963,
        958,
        952,
        947,
        941,
        936,
        931,
        925,
        920,
        914,
        909,
        903,
        898,
        892,
        887,
        882,
        876,
        871,
        865,
        860,
        854,
        849,
        843,
        838,
        832,
        827,
        821,
        816,
        810,
        804,
        799,
        793,
        788,
        782,
        777,
        771,
        766,
        760,
        754,
        749,
        743,
        738,
        732,
        726,
        721,
        715,
        710,
        704,
        698,
        693,
        687,
        681,
        676,
        670,
        664,
        659,
        653,
        647,
        642,
        636,
        630,
        625,
        619,
        613,
        608,
        602,
        596,
        591,
        585,
        579,
        574,
        568,
        562,
        557,
        551,
        545,
        539,
        534,
        528,
        522,
        517,
        511,
        505,
        499,
        494,
        488,
        482,
        476,
        471,
        465,
        459,
        453,
        448,
        442,
        436,
        430,
        425,
        419,
        413,
        407,
        401,
        396,
        390,
        384,
        378,
        372,
        367,
        361,
        355,
        349,
        343,
        337,
        332,
        326,
        320,
        314,
        308,
        302,
        296,
        291,
        285,
        279,
        273,
        267,
        261,
        255,
        249,
        243,
        237,
        231,
        225,
        219,
        213,
        207,
        201,
        195,
        189,
        183,
    ]
}


src/lmt8x/lmt85.py
# datasheet: https://www.ti.com/lit/ds/symlink/lmt85.pdf
transfer_table = {
    'range': [-50, 150],
    'vout': [
        1955,
        1949,
        1942,
        1935,
        1928,
        1921,
        1915,
        1908,
        1900,
        1892,
        1885,
        1877,
        1869,
        1861,
        1853,
        1845,
        1838,
        1830,
        1822,
        1814,
        1806,
        1798,
        1790,
        1783,
        1775,
        1767,
        1759,
        1751,
        1743,
        1735,
        1727,
        1719,
        1711,
        1703,
        1695,
        1687,
        1679,
        1671,
        1663,
        1656,
        1648,
        1639,
        1631,
        1623,
        1615,
        1607,
        1599,
        1591,
        1583,
        1575,
        1567,
        1559,
        1551,
        1543,
        1535,
        1527,
        1519,
        1511,
        1502,
        1494,
        1486,
        1478,
        1470,
        1462,
        1454,
        1446,
        1438,
        1430,
        1421,
        1413,
        1405,
        1397,
        1389,
        1381,
        1373,
        1365,
        1356,
        1348,
        1340,
        1332,
        1324,
        1316,
        1308,
        1299,
        1291,
        1283,
        1275,
        1267,
        1258,
        1250,
        1242,
        1234,
        1225,
        1217,
        1209,
        1201,
        1192,
        1184,
        1176,
        1167,
        1159,
        1151,
        1143,
        1134,
        1126,
        1118,
        1109,
        1101,
        1093,
        1084,
        1076,
        1067,
        1059,
        1051,
        1042,
        1034,
        1025,
        1017,
        1008,
        1000,
        991,
        983,
        974,
        966,
        957,
        949,
        941,
        932,
        924,
        915,
        907,
        898,
        890,
        881,
        873,
        865,
        856,
        848,
        839,
        831,
        822,
        814,
        805,
        797,
        788,
        779,
        771,
        762,
        754,
        745,
        737,
        728,
        720,
        711,
        702,
        694,
        685,
        677,
        668,
        660,
        651,
        642,
        634,
        625,
        617,
        608,
        599,
        591,
        582,
        573,
        565,
        556,
        547,
        539,
        530,
        521,
        513,
        504,
        495,
        487,
        478,
        469,
        460,
        452,
        443,
        434,
        425,
        416,
        408,
        399,
        390,
        381,
        372,
        363,
        354,
        346,
        337,
        328,
        319,
        310,
        301,
    ]
}


src/lmt8x/lmt86.py
# datasheet: https://www.ti.com/lit/ds/symlink/lmt86.pdf
transfer_table = {
    'range': [-50, 150],
    'vout': [
        2616,
        2607,
        2598,
        2589,
        2580,
        2571,
        2562,
        2553,
        2543,
        2533,
        2522,
        2512,
        2501,
        2491,
        2481,
        2470,
        2460,
        2449,
        2439,
        2429,
        2418,
        2408,
        2397,
        2387,
        2376,
        2366,
        2355,
        2345,
        2334,
        2324,
        2313,
        2302,
        2292,
        2281,
        2271,
        2260,
        2250,
        2239,
        2228,
        2218,
        2207,
        2197,
        2186,
        2175,
        2164,
        2154,
        2143,
        2132,
        2122,
        2111,
        2100,
        2089,
        2079,
        2068,
        2057,
        2047,
        2036,
        2025,
        2014,
        2004,
        1993,
        1982,
        1971,
        1961,
        1950,
        1939,
        1928,
        1918,
        1907,
        1896,
        1885,
        1874,
        1864,
        1853,
        1842,
        1831,
        1820,
        1810,
        1799,
        1788,
        1777,
        1766,
        1756,
        1745,
        1734,
        1723,
        1712,
        1701,
        1690,
        1679,
        1668,
        1657,
        1646,
        1635,
        1624,
        1613,
        1602,
        1591,
        1580,
        1569,
        1558,
        1547,
        1536,
        1525,
        1514,
        1503,
        1492,
        1481,
        1470,
        1459,
        1448,
        1436,
        1425,
        1414,
        1403,
        1391,
        1380,
        1369,
        1358,
        1346,
        1335,
        1324,
        1313,
        1301,
        1290,
        1279,
        1268,
        1257,
        1245,
        1234,
        1223,
        1212,
        1201,
        1189,
        1178,
        1167,
        1155,
        1144,
        1133,
        1122,
        1110,
        1099,
        1088,
        1076,
        1065,
        1054,
        1042,
        1031,
        1020,
        1008,
        997,
        986,
        974,
        963,
        951,
        940,
        929,
        917,
        906,
        895,
        883,
        872,
        860,
        849,
        837,
        826,
        814,
        803,
        791,
        780,
        769,
        757,
        745,
        734,
        722,
        711,
        699,
        688,
        676,
        665,
        653,
        642,
        630,
        618,
        607,
        595,
        584,
        572,
        560,
        549,
        537,
        525,
        514,
        502,
        490,
        479,
        467,
        455,
        443,
        432,
        420,
    ]
}


src/lmt8x/__init__.py
__version__ = '1.0.0'


def lmt8x_v2t(model, v):
    if model == 'lmt84':
        from .lmt84 import transfer_table
    elif model == 'lmt85':
        from .lmt85 import transfer_table
    elif model == 'lmt86':
        from .lmt86 import transfer_table
    elif model == 'lmt87':
        from .lmt87 import transfer_table
    else:
        raise ValueError('invalid model "%s"' % model)

    vout = transfer_table['vout']
    start = transfer_table['range'][0]

    if v > vout[0] or v < vout[-1]:
        raise ValueError('v out of range')

    # binary search
    left = 0
    right = len(vout) - 1

    while left < right - 1:
        middle = (left + right) // 2

        # found exact value
        if vout[middle] == v:
            return start + middle

        elif vout[middle] > v:
            left = middle

        else:
            right = middle

    return start + float(vout[left] - v) / float(vout[left] - vout[right]) + left


def lmt84_v2t(v):
    """
    Transfer function for LMT84.

    :param v:  Voltage in mV.
    :type v: float
    :return:  Temperature in Celsius.
    :rtype: float
    """
    return lmt8x_v2t('lmt84', v)


def lmt85_v2t(v):
    """
    Transfer function for LMT85.

    :param v:  Voltage in mV.
    :type v: float
    :return:  Temperature in Celsius.
    :rtype: float
    """
    return lmt8x_v2t('lmt85', v)


def lmt86_v2t(v):
    """
    Transfer function for LMT86.

    :param v:  Voltage in mV.
    :type v: float
    :return:  Temperature in Celsius.
    :rtype: float
    """
    return lmt8x_v2t('lmt86', v)


def lmt87_v2t(v):
    """
    Transfer function for LMT87.

    :param v:  Voltage in mV.
    :type v: float
    :return:  Temperature in Celsius.
    :rtype: float
    """
    return lmt8x_v2t('lmt87', v)


src/lmt8x/lmt87.py
# datasheet: https://www.ti.com/lit/ds/symlink/lmt87.pdf
transfer_table = {
    'range': [-50, 150],
    'vout': [
        3277,
        3266,
        3254,
        3243,
        3232,
        3221,
        3210,
        3199,
        3186,
        3173,
        3160,
        3147,
        3134,
        3121,
        3108,
        3095,
        3082,
        3069,
        3056,
        3043,
        3030,
        3017,
        3004,
        2991,
        2978,
        2965,
        2952,
        2938,
        2925,
        2912,
        2899,
        2886,
        2873,
        2859,
        2846,
        2833,
        2820,
        2807,
        2793,
        2780,
        2767,
        2754,
        2740,
        2727,
        2714,
        2700,
        2687,
        2674,
        2660,
        2647,
        2633,
        2620,
        2607,
        2593,
        2580,
        2567,
        2553,
        2540,
        2527,
        2513,
        2500,
        2486,
        2473,
        2459,
        2446,
        2433,
        2419,
        2406,
        2392,
        2379,
        2365,
        2352,
        2338,
        2325,
        2311,
        2298,
        2285,
        2271,
        2258,
        2244,
        2231,
        2217,
        2204,
        2190,
        2176,
        2163,
        2149,
        2136,
        2122,
        2108,
        2095,
        2081,
        2067,
        2054,
        2040,
        2026,
        2012,
        1999,
        1985,
        1971,
        1958,
        1944,
        1930,
        1916,
        1902,
        1888,
        1875,
        1861,
        1847,
        1833,
        1819,
        1805,
        1791,
        1777,
        1763,
        1749,
        1735,
        1721,
        1707,
        1693,
        1679,
        1665,
        1651,
        1637,
        1623,
        1609,
        1595,
        1581,
        1567,
        1553,
        1539,
        1525,
        1511,
        1497,
        1483,
        1469,
        1455,
        1441,
        1427,
        1413,
        1399,
        1385,
        1371,
        1356,
        1342,
        1328,
        1314,
        1300,
        1286,
        1272,
        1257,
        1243,
        1229,
        1215,
        1201,
        1186,
        1172,
        1158,
        1144,
        1130,
        1115,
        1101,
        1087,
        1073,
        1058,
        1044,
        1030,
        1015,
        1001,
        987,
        973,
        958,
        944,
        929,
        915,
        901,
        886,
        872,
        858,
        843,
        829,
        814,
        800,
        786,
        771,
        757,
        742,
        728,
        713,
        699,
        684,
        670,
        655,
        640,
        626,
        611,
        597,
        582,
        568,
        553,
        538,
    ]
}
